modules/messages/src/lib.rs

```diff
@@ -507,20 +530,57 @@ decl_module! {
 					debug_assert_eq!(message.key.lane_id, lane_id);
 
 					total_messages += 1;
-					if lane.receive_message::<T::MessageDispatch>(relayer_id.clone(), message.key.nonce, message.data) {
-						valid_messages += 1;
-					}
+					let dispatch_weight = T::MessageDispatch::dispatch_weight(&message);
+					let receival_result = lane.receive_message::<T::MessageDispatch, T::AccountId>(
+						&relayer_id_at_bridged_chain,
+						&relayer_id_at_this_chain,
+						message.key.nonce,
+						message.data,
+					);
+
+					// note that we're returning unspent weight to relayer even if message has been
+					// rejected by the lane. This allows relayers to submit spam transactions with
```

Comment:
Why do we do that? I imagine this may cause issues in the future if we for instance want to prioritise bridge transactions or give them a separate (fair) amount of block space (I imagine bridge hub with 3 bridges, were the block weight is split equally between them so they don't have to compete between each other; more sophisticated scheme would also split the unused weight of other bridges).

AFAIU this would lower the cost of a potential griefing attack here. I guess the reasoning is to alleviate excessive costs for multiple concurrent relayers?

Reply:
Yes, this is to lower costs for concurrent relayers. And yes - this may cause a problem if something like that ^^^ would be implemented. But right now it imo solves more problems than it creates - relayer pays the cost of proof parsing, which in case of duplicate messages should be near to 100% of the whole tx cost (duplicated messages are never passed to outside of this module). So the goal is to make post-dispatch weight equal to weight of the same-proof-transaction that declares to bring zero messages.

Reply:
Makes sense, thanks for the explanations!